# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:

Pablo Fenollera Faustino e Iván Villanueva Villalón

Url: https://github.com/Ivavil/VIU-03MIAR-Trabajo_Pra-ctico<br>
Problema:

>3. Combinar cifras y operaciones

Descripción del problema:

Problema 3. Combinar cifras y operaciones

• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)

• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4

• Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo?

(*) La respuesta es obligatoria





                                        

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Respuesta

Partiendo de las restricciones del enunciado en las que se dice que disponemos de las cifras del 1 al 9 y de los 4 signos básicos de las operaciones fundamentales, y sabiendo que no se puede repetir ninguno de ellos, solo se pueden utilizar 5 de los 9 dígitos. Por tanto:
* El número de casos posibles para las cifras es una variacion sin repetición $$V_{n,k} = \frac{n!}{(n-k)!}$$ con n = 9 y k = 5: $$V_{9,5} = \frac{9!}{(9-5)!} = \frac{9!}{4!} = 9 \times 8 \times 7 \times 6 \times 5 = 15120$$

* El número de casos disponibles para las operaciones fundamentales es una permutación sin repetición $$P_n = n!$$ con n = 4: 

$$P_4 = 4! = 4 \times 3 \times 2 \times 1 = 24$$



Al tratarse de sucesos independientes, ya que a la hora de generar las expresiones, números y operaciones no dependen los unos de los otros, el número total de posibles expresiones es: $$casos_{totales} =  15120 \times 24 = 362880$$

Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Respuesta

Se ha decidido guardar los datos de entrada iniciales, es decir, los dígitos y las operaciones en 2 listas separadas. La razón detrás de esto es que en el algoritmo recorremos de manera iterativa los elementos para generar las posibles expresiones y no requerimos de estructuras especializadas.

In [1]:
# numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
# symbols = ['+','-','*','/']

Para guardar los resultados de las expresiones, se utiliza un conjunto (set) para evitar valores duplicados, ya que solo interesa registrar que es posible obtener los valores y no las veces que se repiten

In [2]:
# final_results = set()

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Respuesta

La función objetivo genera expresiones válidas que produzcan un número entero, no se busca maximizar ni minimizar un valor numérico.
Es un problema de búsqueda exhaustiva en un espacio de soluciones.

Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

In [11]:
def vars_without_rep(elems, k):
    global count
    # Se filtra el caso de que k sea mayor que el número de elementos, que no es posible
    if k > len(elems):
        return []

    results = []

    def backtrack(current_var, elems_availables):
        global count
        # Se guarda la variación actual si ya tiene el tamaño k
        if len(current_var) == k:
            results.append(current_var)
            return

        # Iteración sobre los elementos disponibles
        for i in range(len(elems_availables)):
            elem = elems_availables[i]
            
            # Nuevos elementos = todos - elección actual
            new_available = elems_availables[:i] + elems_availables[i+1:]
            
            # Llamada recursiva añadiendo el elemento actual
            count += 1
            backtrack(current_var + [elem], new_available)

    # Se inicia la recursión con una lista vacía y todos los elementos disponibles
    backtrack([], elems)
    print
    return results

def perms_without_rep(elems):
    global count
    results = []

    def backtrack(init):
        global count
        if init == len(elems):
            # Copia de la lista actual, ya que la original seguirá cambiando en los siguientes pasos.
            results.append(elems.copy())
            return

        # Iteración desde el índice 'init' hasta el final
        for i in range(init, len(elems)):
            # Se intercambia el elemento en 'init' con el elemento en 'i'
            elems[init], elems[i] = elems[i], elems[init]
            count += 1
            
            # Llamada recursiva para permutar el resto de la lista
            backtrack(init + 1)
            
            # Se deshace el intercambio para restaurar el orden original y poder explorar las siguientes ramas
            elems[init], elems[i] = elems[i], elems[init]

    # Recursión desde el índice 0
    backtrack(0)
    
    return results

def brute_force():
    global count
    numbers = ['1','2','3','4','5','6','7','8','9']
    symbols = ['+','-','*','/']
    
    # Se obtienen todas las variaciones posibles de grupos de 5 números extraídos del conjunto de 9 números
    var_numbers = vars_without_rep(numbers,5)
    
    # Se obtienen todas las permutaciones posibles de los símbolos a utilizar
    perm_symbols = perms_without_rep(symbols)
    expressions = []
    final_results=set()
    for perm_sym in perm_symbols:
        for var_num in var_numbers:
            # Generación de todas las combinaciones posibles
            expression = var_num[0]+perm_sym[0]+var_num[1]+perm_sym[1]+var_num[2]+perm_sym[2]+var_num[3]+perm_sym[3]+var_num[4]
            result = eval(expression)
            count += 1
            # Si tras evaluar la expresión el número es entero lo añadimos a la lista de evals
            if result%1==0:
                final_results.add(result)
            expressions.append(expression)

    print(f"Valor Máximo: {max(final_results)}")
    print(f"Valor Mínimo: {min(final_results)}")
    print(f"Resultados enteros obtenidos: {sorted(list(final_results))}")
    print("El numero total de operaciones es " + str(count))

count = 0
brute_force()

Valor Máximo: 77.0
Valor Mínimo: -69.0
Resultados enteros obtenidos: [-69.0, -68.0, -67.0, -66.0, -65.0, -64.0, -63.0, -62.0, -61.0, -60.0, -59.0, -58.0, -57.0, -56.0, -55.0, -54.0, -53.0, -52.0, -51.0, -50.0, -49.0, -48.0, -47.0, -46.0, -45.0, -44.0, -43.0, -42.0, -41.0, -40.0, -39.0, -38.0, -37.0, -36.0, -35.0, -34.0, -33.0, -32.0, -31.0, -30.0, -29.0, -28.0, -27.0, -26.0, -25.0, -24.0, -23.0, -22.0, -21.0, -20.0, -19.0, -18.0, -17.0, -16.0, -15.0, -14.0, -13.0, -12.0, -11.0, -10.0, -9.0, -8.0, -7.0, -6.0, -5.0, -4.0, -3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0, 31.0, 32.0, 33.0, 34.0, 35.0, 36.0, 37.0, 38.0, 39.0, 40.0, 41.0, 42.0, 43.0, 44.0, 45.0, 46.0, 47.0, 48.0, 49.0, 50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 63.0, 64.0, 65.0, 66.0, 67.0, 68.0, 69.0, 70.0, 71.0, 72.0, 73.0, 74.0, 75.0, 76.0, 77.0

Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

El algoritmo de fuerza bruta es de complejidad factorial porque la propia estructura del problema te obliga a generar permutaciones y variaciones, y las dos crecen factorialmente (aunque las variaciones no crecen de manera "puramente" factorial). Por tanto la complejidad es

$$O(n!\times m!)$$

Nota: en nuestro caso específico m representa los datos de entrada de las operaciones y estas tienen un tamaño finito. Si suponemos que el tamaño de m va a ser siempre 4 (por las operaciones fundamentales), entonces su peso a la hora de calcular la complejidad computacional del algoritmo podria despreciarse de modo que 

$$O(n!)$$

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

In [17]:
def bitmask_dp_hashmap():
    global count
    # Cada valor del diccionario (cuyas claves son de 0 a 4) guarda los cálculos dependiendo de cuántos signos matemáticos
    # se han usado hasta el momento.
    dp = {k: {} for k in range(len(symbols) + 1)}
    
    # Función auxiliar que aplica el siguiente operador matemático (+, -, *, /) a la expresión de entrada
    def apply_operator(states, op_idx, next_num):
        global count
        new_states = set()
        for total, last_term in states:
            count += 1
            match op_idx:
                case 0: # Suma
                    new_states.add((total + next_num, next_num))
                case 1: # Resta
                    new_states.add((total - next_num, -next_num))
                case 2: # Multiplicación
                    new_term = last_term * next_num
                    new_states.add((total - last_term + new_term, new_term))
                case 3: # División
                    if next_num != 0 and last_term % next_num == 0:
                        new_term = last_term // next_num
                        new_states.add((total - last_term + new_term, new_term))
        return new_states

    # Se elige un número inicial al azar
    for i, num in enumerate(numbers):
        count += 1
        dp[0][(1 << i, 0)] = {(num, num)}
        
    # En cada paso del bucle se van añadiendo los operadores de uno en uno.
    for k in range(len(symbols)):
        for (num_mask, op_mask), current_values in dp[k].items():
            
            # Separación de los números y los signos que aún están libres para usarse.
            valid_nums = [(i, numbers[i]) for i in range(len(numbers)) if not (num_mask & (1 << i))]
            valid_ops = [op for op in range(len(symbols)) if not (op_mask & (1 << op))]
            
            # Exploración de todos los caminos posibles mezclando los que todavía están libres.
            for i, next_num in valid_nums:
                new_num_mask = num_mask | (1 << i)
                
                for op_idx in valid_ops:
                    new_op_mask = op_mask | (1 << op_idx)
                    count += 1
                    # Se guarda el nuevo resultado para el siguiente turno.
                    new_values = apply_operator(current_values, op_idx, next_num)
                    
                    if new_values:
                        state_key = (new_num_mask, new_op_mask)
                        dp[k + 1].setdefault(state_key, set()).update(new_values)

    # Revisión del último elemento del diccionario (donde ya se usan los 4 signos) para asegurar de 
    # que la combinación tiene exactamente 5 números elegidos.
    final_results = set()
    for (num_mask, _), values in dp[len(symbols)].items():
        if num_mask.bit_count() == 5:
            for total, _ in values:
                count += 1
                final_results.add(total)

    print(f"Total de resultados únicos enteros: {len(final_results)}")
    if final_results:
        print(f"Valor Máximo: {max(final_results)}")
        print(f"Valor Mínimo: {min(final_results)}")
        print(f"Resultados enteros obtenidos: {sorted(list(final_results))}")
        print("El numero total de operaciones es " + str(count))

if __name__ == "__main__":
    count = 0
    numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    symbols = ['+','-','*','/']
    bitmask_dp_hashmap()

Total de resultados únicos enteros: 147
Valor Máximo: 77
Valor Mínimo: -69
Resultados enteros obtenidos: [-69, -68, -67, -66, -65, -64, -63, -62, -61, -60, -59, -58, -57, -56, -55, -54, -53, -52, -51, -50, -49, -48, -47, -46, -45, -44, -43, -42, -41, -40, -39, -38, -37, -36, -35, -34, -33, -32, -31, -30, -29, -28, -27, -26, -25, -24, -23, -22, -21, -20, -19, -18, -17, -16, -15, -14, -13, -12, -11, -10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77]
El numero total de operaciones es 97874


El algoritmo anterior mejora la complejidad del algoritmo de fuerza bruta porque implementa los siguientes mecanismos:

1. Eliminación de Redundancias (Programación Dinámica): El algoritmo resuelve y almacena en memoria los subproblemas matemáticos una única vez. Al reutilizar estos estados precalculados, erradica la repetición masiva de cálculos idénticos inherente a la fuerza bruta.
  
2. Poda del Espacio de Búsqueda: Interrumpe la ejecución de una rama computacional en el instante en que detecta una operación que viola las restricciones del problema (como generar números racionales). Esto evita el procesamiento inútil de todas las permutaciones descendentes derivadas de ese estado inválido.
  
3. Reducción de la Complejidad Asintótica: Al utilizar Máscaras de Bits, transforma la representación del problema: pasa de evaluar permutaciones dependientes del orden (crecimiento factorial $O(n!)$) a evaluar combinaciones de conjuntos (crecimiento exponencial $O(2^n)$), reduciendo matemáticamente la escala global de iteraciones.

Además, se puede revisar que el número de operaciones (en este caso, se simplica contando únicamente las iteraciones dentro de bucles, sin contar individualmente cada operación) del algoritmo de fuerza bruta es 381673, mientras que el número de operaciones del algoritmo optimizado es 97874. Por tanto, el cambio de algoritmo ha reducido el número de operaciones un $$ 1 - \left( \frac{97874}{381673} \right) \cdot 100 = 74\%$$

(*)Calcula la complejidad del algoritmo

Respuesta

Como el algoritmo itera sobre todos los subconjuntos posibles de los $n$ números disponibles, es decir, los $2^n$ estados de las máscara de bits con las que se trabaja en este algoritmo, la complejidad es $O(2^n \cdot 2^m)$.

Nota: Como en la justificación de la complejidad del algoritmo de fuerza bruta, si suponemos que el tamaño de m va a ser siempre 4 (por las operaciones fundamentales), entonces su peso a la hora de calcular la complejidad computacional del algoritmo podria despreciarse de modo que la complejidad resultante se podría aproximar a $O(2^n)$.

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

In [ ]:
import random

symbols = ['+','-','*','/']
numbers = random.sample(range(1, 20), 9)
print(numbers)

[13, 10, 18, 3, 2, 15, 1, 16, 11]


Este problema específico está diseñado para funcionar de manera asegurada para los números del 1 al 9, y las operaciones fundamentales. Por lo que no es posible asegurar que al generar un juego de datos de entrada aleatorios de todos los puntos del problema que el algoritmo debería resolver se cumplan. En concreto, el de comprobar si es posible obtener todos los enteros entre el máximo y el mínimo a partir de los datos de entrada.

Por razones obvias, no se va a modificar el conjunto de datos de las operaciones fundamentales, ya que habría que modificar el algoritmo para incluir las nuevas operaciones.

Aplica el algoritmo al juego de datos generado

Respuesta

In [ ]:
bitmask_dp_hashmap()

Total de resultados únicos enteros: 557
Valor Máximo: 301
Valor Mínimo: -283
Resultados enteros obtenidos: [-283, -282, -281, -280, -276, -275, -274, -273, -272, -271, -270, -268, -267, -265, -264, -263, -262, -261, -260, -259, -258, -257, -256, -255, -254, -252, -251, -249, -247, -246, -244, -243, -241, -235, -234, -233, -232, -230, -229, -228, -227, -226, -225, -224, -223, -222, -221, -220, -219, -218, -217, -216, -215, -214, -213, -212, -211, -209, -208, -207, -203, -202, -201, -200, -198, -196, -195, -194, -193, -192, -191, -190, -189, -188, -187, -186, -185, -184, -183, -182, -181, -180, -179, -178, -177, -176, -175, -174, -173, -172, -171, -170, -169, -168, -167, -166, -165, -164, -163, -162, -161, -160, -159, -158, -157, -156, -155, -154, -153, -152, -151, -150, -149, -148, -147, -146, -145, -144, -143, -142, -141, -140, -139, -138, -137, -136, -135, -134, -133, -132, -131, -130, -129, -128, -127, -126, -125, -124, -123, -122, -121, -120, -119, -118, -117, -116, -115, -114, -113

Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

Se ha utilizado la inteligencia artificial, en concreto Gemini 3.1 Pro, para ayudar en algunos puntos de la generación del algoritmo de fuerza bruta y la generación del algoritmo de programaciónn dinámica

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

El algoritmo ya está diseñado para funcionar con cualquier tamaño o valor de los datos de entrada numéricos (otra cosa es que se cumpla la condición de la existencia del resto de enteros entre el máximo y el mínimo).  Una posible ampliación del problema y, por tanto, del algoritmo, sería ampliar el número de operaciones matemáticas disponibles a realizar. También se podría añadir paréntesis en las expresiones para darle prioridad a unas operaciones sobre otras, de modo que el espacio de posibilidades aumentaría drásticamente.